### PR TITLE
Fix all null list column with missing child column in JSON reader

### PR DESCRIPTION
**Pull request by karthikeyann**
Fixes https://github.com/rapidsai/cudf/issues/17341 and fixes https://github.com/rapidsai/cudf/issues/17349.
- **expose make_empty_column**
- **fix all null list with empty nested child on missing list child column when prune columns is enabled**
- **add unit test**

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/io/json/json_column.cu [hunk 1]
```diff
@@ -464,26 +464,37 @@ std::pair<std::unique_ptr<column>, std::vector<column_name_info>> device_json_co
...
         json_col.child_columns.empty() ? list_child_name : json_col.child_columns.begin()->first);
 
-      // Note: json_col modified here, reuse the memory
-      auto offsets_column = std::make_unique<column>(data_type{type_id::INT32},
-                                                     num_rows + 1,
-                                                     json_col.child_offsets.release(),
-                                                     rmm::device_buffer{},
-                                                     0);
+      // if child is not present, set the null mask correctly, but offsets are zero, and children
+      // are empty Note: json_col modified here, reuse the memory
```
Comment (ttnghia) [inline cpp/src/io/json/json_column.cu:468]:
```suggestion
      // If child is not present, set the null mask correctly, but offsets are zero, and children are empty.
      // Note: json_col modified here, reuse the memory
```

File: cpp/src/io/json/parser_features.cpp [hunk 2]
```diff
@@ -88,7 +88,15 @@ struct empty_column_functor {
...
-    return make_lists_column(0, std::move(offsets), std::move(child), 0, {}, stream, mr);
+    std::vector<std::unique_ptr<column>> child_columns;
+    child_columns.push_back(std::move(offsets));
+    child_columns.push_back(std::move(child));
+    return std::make_unique<column>(cudf::data_type{type_id::LIST},
+                                    0,
+                                    rmm::device_buffer{},
+                                    rmm::device_buffer{},
+                                    0,
+                                    std::move(child_columns));
```
Comment (davidwendt) [inline cpp/src/io/json/parser_features.cpp:99]: Why is this change needed?
Could this utility be used instead? https://github.com/rapidsai/cudf/blob/9cc907122077d18e5128e7da36685fdeb82fef41/cpp/include/cudf/lists/detail/lists_column_factories.hpp#L50-L52

File: cpp/src/io/json/parser_features.cpp [hunk 3]
```diff
@@ -99,10 +107,22 @@ struct empty_column_functor {
       child_columns.push_back(cudf::type_dispatcher(
         schema.child_types.at(child_name).type, *this, schema.child_types.at(child_name)));
     }
-    return make_structs_column(0, std::move(child_columns), 0, {}, stream, mr);
+    return std::make_unique<column>(cudf::data_type{type_id::STRUCT},
+                                    0,
+                                    rmm::device_buffer{},
+                                    rmm::device_buffer{},
+                                    0,
+                                    std::move(child_columns));
```
Comment (davidwendt) [inline cpp/src/io/json/parser_features.cpp:115]: What does this do different than the factory call?

**Review by davidwendt — CHANGES_REQUESTED** [review]

File: cpp/src/io/json/parser_features.cpp (hunk 3 above)
Comment (ttnghia) [reply cpp/src/io/json/parser_features.cpp:115]: We discovered through profiling that `make_lists_column` and `make_structs_column` introduce significant overhead by unnecessarily calling to `purge_non_empty_nulls` and `superimpose_nulls_no_sanitize`. These sanitization may be needed in general, however, here we know that the child(ren) of JSON output column cannot have non-null rows if the parent is null (due to invalid input) thus we want to avoid such overhead by calling to the column constructor directly.

Comment (ttnghia) [reply cpp/src/io/json/parser_features.cpp:115]: @karthikeyann Please add comments to clarify why:
```
// Do not use `cudf::make_lists_column` since we do not need to call `purge_nonempty_nulls`
// on the child column as it does not have non-empty nulls.
```
and
```
// Do not use `cudf::make_structs_column` since we do not need to call `superimpose_nulls`
// on the children columns.
```

Comment (ttnghia) [reply cpp/src/io/json/parser_features.cpp:115]: I've filed a related issue for this: https://github.com/NVIDIA/spark-rapids-jni/issues/2602

Comment (davidwendt) [reply cpp/src/io/json/parser_features.cpp:115]: https://github.com/rapidsai/cudf/issues/12567

Comment (davidwendt) [reply cpp/src/io/json/parser_features.cpp:115]: Seems we could fix the `make_structs_column` and `make_lists_column` to not call these utilities for empty columns. 
Or the utilities could be fixed to no-op for empty columns perhaps?

File: cpp/tests/io/json/json_test.cpp [hunk 4]
```diff
@@ -3250,6 +3250,62 @@ TEST_F(JsonReaderTest, JsonNestedDtypeFilterWithOrder)
...
+    EXPECT_EQ(result.metadata.schema_info[1].children[1].name, "element");
+    ASSERT_EQ(result.metadata.schema_info[1].children[1].children.size(), 1);
+    EXPECT_EQ(result.metadata.schema_info[1].children[1].children[0].name, "d");
+    EXPECT_EQ(result.tbl->get_column(1).type().id(), cudf::type_id::LIST);
+    EXPECT_EQ(result.tbl->get_column(1).child(0).type().id(), cudf::type_id::INT32);
+    EXPECT_EQ(result.tbl->get_column(1).child(1).type().id(), cudf::type_id::STRUCT);
+    EXPECT_EQ(result.tbl->get_column(1).child(1).child(0).type().id(), cudf::type_id::STRING);
+    EXPECT_EQ(result.tbl->get_column(1).size(), 2);
+    EXPECT_EQ(result.tbl->get_column(1).child(1).size(), 0);
+    EXPECT_EQ(result.tbl->get_column(1).child(1).child(0).size(), 0);
```
Comment (ttnghia) [inline cpp/tests/io/json/json_test.cpp:3307]:
```suggestion
  auto const expected0 = [&] {
    auto const get_structs = [] {
        auto child = cudf::test::fixed_width_column_wrapper<int64_t>{1};
        return cudf::test::structs_column_wrapper{{child}};
      };
      auto const valids = std::vector<bool>{1, 0};
      auto [null_mask, null_count] = cudf::test::detail::make_null_mask(valids.begin(), valids.end());
      return cudf::make_lists_column(
        2, cudf::test::fixed_width_column_wrapper<cudf::size_type>{0, 1, 1}.release(), get_structs().release(), null_count, std::move(null_mask));
  }();
  
    auto const expected0 = [&] {
    auto const get_structs = [] {
        auto child = cudf::test::strings_column_wrapper{};
        return cudf::test::structs_column_wrapper{{child}};
      };
      auto const valids = std::vector<bool>{0, 0};
      auto [null_mask, null_count] = cudf::test::detail::make_null_mask(valids.begin(), valids.end());
      return cudf::make_lists_column(
        2, cudf::test::fixed_width_column_wrapper<cudf::size_type>{0, 0, 0}.release(), get_structs().release(), null_count, std::move(null_mask));
  }();
  
  CUDF_TEST_EXPECT_COLUMNS_EQUAL(*expected0, result.tbl->get_column(0).view());
  CUDF_TEST_EXPECT_COLUMNS_EQUAL(*expected1, result.tbl->get_column(1).view());
```

File: cpp/src/io/json/parser_features.cpp (hunk 3 above)
Comment (karthikeyann) [reply cpp/src/io/json/parser_features.cpp:119]: Either fix is good.

File: cpp/tests/io/json/json_test.cpp [hunk 5]
```diff
@@ -3253,6 +3254,84 @@ TEST_F(JsonReaderTest, JsonNestedDtypeFilterWithOrder)
...
+            {data_type{cudf::type_id::STRUCT},
+             {
+               {"d", {data_type{cudf::type_id::STRING}}},
+             },
+             {{"d"}}}}}}},
+      },
+      {{"a", "c2"}}};
+    in_options.set_dtypes(dtype_schema);
+    cudf::io::table_with_metadata result = cudf::io::read_json(in_options);
+    // Make sure we have column "a":[float]
```
Comment (ttnghia) [inline cpp/tests/io/json/json_test.cpp:3287]:
```suggestion
    // Make sure we have column "a":[int64_t]
```

File: cpp/tests/io/json/json_test.cpp [hunk 6]
```diff
@@ -3253,6 +3254,84 @@ TEST_F(JsonReaderTest, JsonNestedDtypeFilterWithOrder)
...
+    ASSERT_EQ(result.tbl->num_columns(), 2);
+    ASSERT_EQ(result.metadata.schema_info.size(), 2);
+    EXPECT_EQ(result.metadata.schema_info[0].name, "a");
+    ASSERT_EQ(result.metadata.schema_info[0].children.size(), 2);
+    EXPECT_EQ(result.metadata.schema_info[0].children[0].name, "offsets");
+    EXPECT_EQ(result.metadata.schema_info[0].children[1].name, "element");
+    // Make sure we have all null list "c2": [{"d": ""}]
```
Comment (ttnghia) [inline cpp/tests/io/json/json_test.cpp:3294]: Test fails because the output for this column is still not all-null:
```
cudf::list_view<cudf::struct_view<cudf::string_view,>>:
Length : 2
Offsets : 0, 0, 0
Null count: 1
01
   cudf::struct_view<cudf::string_view,>:
   Length : 0:
```

File: cpp/src/io/json/json_column.cu [hunk 7]
```diff
@@ -509,17 +509,18 @@ std::pair<std::unique_ptr<column>, std::vector<column_name_info>> device_json_co
...
       column_names.back().children      = names;
       auto [result_bitmask, null_count] = make_validity(json_col);
-      auto ret_col                      = make_lists_column(num_rows,
-                                       std::move(offsets_column),
-                                       std::move(child_column),
-                                       0,
-                                       rmm::device_buffer{0, stream, mr},
-                                       stream,
-                                       mr);
+      auto ret_col                      = make_lists_column(
```
[the remark this answers is not in the document]
Comment (karthikeyann) [reply cpp/src/io/json/json_column.cu:512]: Here, purge non-empty nulls is required when prune columns is enabled.

File: cpp/src/io/json/parser_features.cpp (hunk 2 above)
Comment (karthikeyann) [reply cpp/src/io/json/parser_features.cpp:101]: It works only for fixed width type child.

Comment (karthikeyann) [reply cpp/src/io/json/parser_features.cpp:101]: This change is to avoid any `purge/has_nonempty_nulls` calls.

File: cpp/src/io/json/host_tree_algorithms.cu [hunk 8]
```diff
@@ -794,6 +807,14 @@ std::pair<cudf::detail::host_vector<bool>, hashmap_of_device_columns> build_tree
                      "list child column insertion failed, duplicate column name in the parent");
         ref.get().column_order.emplace_back(list_child_name);
         auto this_ref = std::ref(ref.get().child_columns.at(list_child_name));
+        if (options.is_enabled_experimental()) {
+          for (auto const& child_id : child_ids) {
+            if (is_pruned[child_id])
+              std::cout << "pruned.child " << column_names[root] << std::endl;
```
Comment (lamarrr) [inline cpp/src/io/json/host_tree_algorithms.cu:813]: debug artifact?

File: cpp/src/io/json/host_tree_algorithms.cu [hunk 9]
```diff
@@ -921,9 +945,33 @@ void scatter_offsets(tree_meta_t const& tree,
...
+     row_offsets       = row_offsets.begin(),
+     d_is_mixed_pruned = d_is_mixed_pruned.begin(),
+     d_ignore_vals     = d_ignore_vals.begin(),
+     d_columns_data    = d_columns_data.begin()] __device__(size_type i) {
+      auto const node_id        = node_ids[i];
+      auto const parent_node_id = parent_node_ids[node_id];
+      if (parent_node_id == parent_node_sentinel or d_ignore_vals[col_ids[parent_node_id]]) return;
+      if (column_categories[col_ids[parent_node_id]] == NC_LIST and
+          d_is_mixed_pruned[col_ids[node_id]]) {
+        // printf("clearing bit %d %d\n", parent_node_id, row_offsets[parent_node_id]);
```
Comment (lamarrr) [inline cpp/src/io/json/host_tree_algorithms.cu:967]: debug artifact

File: cpp/tests/io/json/json_test.cpp [hunk 10]
```diff
@@ -3253,6 +3255,144 @@ TEST_F(JsonReaderTest, JsonNestedDtypeFilterWithOrder)
...
+    CUDF_TEST_EXPECT_COLUMNS_EQUAL(*expected0, result.tbl->get_column(0).view());
+    CUDF_TEST_EXPECT_COLUMNS_EQUAL(*expected1, result.tbl->get_column(1).view());
+  }
+}
+
+TEST_F(JsonReaderTest, NullifyMixedList)
+{
+  using namespace cudf::test::iterators;
+  // test list
+  std::string json_stringl = R"(
```
Comment (shrshi) [inline cpp/tests/io/json/json_test.cpp:3339]: In this example, if we have an input such as `{"c2": [{"b": [1], "c": 123}]}`, would the expected output be `[{null, null}]`?

File: cpp/src/io/json/json_column.cu [hunk 11]
```diff
@@ -464,46 +464,51 @@ std::pair<std::unique_ptr<column>, std::vector<column_name_info>> device_json_co
...
+        std::move(child_column),
+        null_count,
+        null_count == 0 ? rmm::device_buffer{0, stream, mr} : std::move(result_bitmask),
+        stream,
+        mr);
       // The null_mask is set after creation of list column is to skip the purge_nonempty_nulls and
       // null validation applied in make_lists_column factory, which is not needed for json
       // parent column cannot be null when its children is non-empty in JSON
-      if (null_count != 0) { ret_col->set_null_mask(std::move(result_bitmask), null_count); }
+      // if (null_count != 0) { ret_col->set_null_mask(std::move(result_bitmask), null_count); }
```
Comment (shrshi) [inline cpp/src/io/json/json_column.cu:511]: This line can be removed

File: cpp/tests/io/json/json_test.cpp (hunk 10 above)
Comment (ttnghia) [reply cpp/tests/io/json/json_test.cpp:3339]: Spark CPU shows this:
```
scala> val df = Seq("""{"c2": [{"b": [1], "c": 123}]}""").toDF
df: org.apache.spark.sql.DataFrame = [value: string]

scala> df.repartition(1).selectExpr("from_json(value, 'struct<c2: array<struct<b: string, c: string>>>')").show(false)
+----------------+
|from_json(value)|
+----------------+
|{[{[1], 123}]}  |
+----------------+
```
So the output should be `[{[1], 123}]`. And this PR outputs that result :+1:

Comment (ttnghia) [reply cpp/tests/io/json/json_test.cpp:3339]: Note that the first value is `[1]` is a string, not an array of string/number:
```
scala> df.repartition(1).selectExpr("from_json(value, 'struct<c2: array<struct<b: string, c: string>>>')").printSchema
root
 |-- from_json(value): struct (nullable = true)
 |    |-- c2: array (nullable = true)
 |    |    |-- element: struct (containsNull = true)
 |    |    |    |-- b: string (nullable = true)
 |    |    |    |-- c: string (nullable = true)
```
So this is due to read mixed type as string.

Comment (ttnghia) [reply cpp/tests/io/json/json_test.cpp:3339]: For mismatched nested schema, the second value is still not null:
```

scala> val df = Seq("""{"c2": [{"b": [1], "c": 123}]}""").toDF
df: org.apache.spark.sql.DataFrame = [value: string]

scala> df.show(false)
+------------------------------+
|value                         |
+------------------------------+
|{"c2": [{"b": [1], "c": 123}]}|
+------------------------------+


scala> df.repartition(1).selectExpr("from_json(value, 'struct<c2: array<struct<b: struct<d:int>, c: string>>>')").show(false)
+----------------+
|from_json(value)|
+----------------+
|{[{null, 123}]} |
+----------------+
```

File: cpp/src/io/json/json_column.cu [hunk 12]
```diff
@@ -464,46 +464,51 @@ std::pair<std::unique_ptr<column>, std::vector<column_name_info>> device_json_co
...
-                                              prune_columns,
-                                              child_schema_element,
-                                              stream,
-                                              mr);
+      auto child_schema_element  = get_list_child_schema();
+      auto [child_column, names] = [&]() {
+        if (json_col.child_columns.empty()) {
+          // EMPTY type could not used because gather throws exception on EMPTY type.
+          auto empty_col = make_empty_column(
+            child_schema_element.value_or(schema_element{data_type{type_id::INT16}}), stream, mr);
```
Comment (ttnghia) [inline cpp/src/io/json/json_column.cu:480]: Why it is `INT16` now?

File: cpp/src/io/json/parser_features.cpp (hunk 3 above)
Comment (ttnghia) [reply cpp/src/io/json/parser_features.cpp:119]: Ah yes, I was wrong on this: it is unnecessary here since the input column is empty.

File: cpp/tests/io/json/json_test.cpp (hunk 10 above)
Comment (karthikeyann) [reply cpp/tests/io/json/json_test.cpp:3339]: Even with mixed_type_as_string=false, specifying a schema as string, forces all its rows to be returned as string (even if it has list/struct/string/fixed_width_types). This came from one of requirements from spark.

File: cpp/src/io/json/json_column.cu (hunk 12 above)
Comment (karthikeyann) [reply cpp/src/io/json/json_column.cu:480]: That was for debugging. I replaced it with INT8. 
`INT8` is still a placeholder because we can not create a empty column in libcudf.

File: cpp/src/io/json/host_tree_algorithms.cu (hunk 8 above)
Comment (karthikeyann) [reply cpp/src/io/json/host_tree_algorithms.cu:813]: removed.

**Review by ttnghia — APPROVED** [review]

**Review by lamarrr — APPROVED** [review]
Looks  good to me

File: cpp/src/io/json/host_tree_algorithms.cu [hunk 13]
```diff
@@ -921,9 +945,32 @@ void scatter_offsets(tree_meta_t const& tree,
              column_categories[col_ids[parent_node_id]] == NC_LIST and
              (!d_ignore_vals[col_ids[parent_node_id]]);
     });
+  // For children of list and in ignore_vals, find it's parent node id, and set corresponding
+  // parent's null mask to null. Setting mixed type list rows to null.
+  thrust::for_each_n(
+    rmm::exec_policy_nosync(stream),
+    thrust::make_counting_iterator<size_type>(0),
+    list_children_end - thrust::make_zip_iterator(node_ids.begin(), parent_col_ids.begin()),
```
Comment (davidwendt) [inline cpp/src/io/json/host_tree_algorithms.cu:953]: Should this use `thrust::distance` ?
```suggestion
    thrust::distance(thrust::make_zip_iterator(node_ids.begin(), parent_col_ids.begin(), list_children_end),
```

File: cpp/src/io/json/parser_features.cpp [hunk 14]
```diff
@@ -99,10 +109,24 @@ struct empty_column_functor {
...
     }
-    return make_structs_column(0, std::move(child_columns), 0, {}, stream, mr);
+    // Do not use `cudf::make_structs_column` since we do not need to call `superimpose_nulls` on
+    // the children columns. Look issue #17356
+    return std::make_unique<column>(cudf::data_type{type_id::STRUCT},
+                                    0,
+                                    rmm::device_buffer{},
+                                    rmm::device_buffer{},
+                                    0,
+                                    std::move(child_columns));
```
Comment (karthikeyann) [inline cpp/src/io/json/parser_features.cpp:119]:
```suggestion
    return make_structs_column(0, std::move(child_columns), 0, {}, stream, mr);

```

**Review by davidwendt — APPROVED** [review]

File: cpp/src/io/json/host_tree_algorithms.cu (hunk 13 above)
Comment (ttnghia) [reply cpp/src/io/json/host_tree_algorithms.cu]: The same expression is used to compute `num_list_children` below this line.

Comment (ttnghia) [reply cpp/src/io/json/host_tree_algorithms.cu]: Should be something like this:
```
auto const list_children_begin = thrust::make_zip_iterator(node_ids.begin(), parent_col_ids.begin();
// replace all `thrust::make_zip_iterator` by `list_children_begin`.
auto const num_list_children = thrust::distance(list_children_begin, list_children_end);
// use num_list_children
```

File: cpp/src/io/json/host_tree_algorithms.cu [hunk 15]
```diff
@@ -947,10 +947,12 @@ void scatter_offsets(tree_meta_t const& tree,
     });
   // For children of list and in ignore_vals, find it's parent node id, and set corresponding
   // parent's null mask to null. Setting mixed type list rows to null.
+  auto const num_list_children = thrust::distance(
+    thrust::make_zip_iterator(node_ids.begin(), parent_col_ids.begin()), list_children_end);
```
Comment (ttnghia) [inline cpp/src/io/json/host_tree_algorithms.cu:951]: Please remove lines 973-975.
